### DISCUSSION TITLE
How do I know what's wrong?

### DISCUSSION THREAD

**Discussion by leex2000** [nlohmann/json#3771] [Q&A]
If I use the pattern with exception, i can get error info from ex.what()
```
json jsonContent = json::parse(content);
try
{
	jsonContent = json::parse(content);
}
catch (const nlohmann::detail::exception& ex)
{
	cout << ex.what() << endl;
}
```

I really don't like try...throw, I choose to use a pattern without exceptions.
```
json jsonContent = json::parse(content, nullptr, false);
if (jsonContent.is_discarded() || !jsonContent.is_object())
{
	// How do I know what's wrong?
}
```

**Comment by falbrechtskirchinger** [comment]
There's no easy way.

It can be done using the [`json::sax_parse()`](https://json.nlohmann.me/api/basic_json/sax_parse/) function and intercepting the [parse error](https://json.nlohmann.me/api/json_sax/parse_error/).

**Reply by falbrechtskirchinger** [reply]
If you do go this route, I suggest you copy the `nlohmann::details::json_sax_dom_parser` class into your codebase and modify it. You'll have to remove the `JSON_HEDLEY_*` macros because they are undefined at the end of `json.hpp`

**Reply by leex2000** [reply]
> There's no easy way.
> 
> It can be done using the [`json::sax_parse()`](https://json.nlohmann.me/api/basic_json/sax_parse/) function and intercepting the [parse error](https://json.nlohmann.me/api/json_sax/parse_error/).

Thank for answer. It's too much trouble for me.
I already had to use the pattern with exception. That makes my code clumsy.
```
bool MyFunction()
{
	// ...

	try
	{
		jsonContent = json::parse(content);
	}
	catch (const nlohmann::detail::exception& ex)
	{
		cout << ex.what() << endl;
		// some cleanup code
		return false;
	}

	if (!jsonContent.is_object())
	{
		cout << "value is't an object" << endl;
		// some cleanup code
		return false;
	}

	// do something
}
```
That's why I don't like try...catch. 
I hope there will be a version that doesn't rely on exceptions.

**Reply by falbrechtskirchinger** [reply]
Removing exceptions is a tall order – speaking from experience. It's easy enough to remove exceptions from the parser, but a lot more complicated to remove them from the rest of the library without breaking changes. C++11 compatibility is what ultimately made me give up on an upstreamable solution.
